api/src/main/java/jakarta/servlet/ServletContext.java

```diff
@@ -1436,4 +1523,97 @@
+     * <li>{@link ServletContext#getContext(String)}</li>
+     * <li>{@link ServletContext#getRequestDispatcher(String)}</li>
+     * <li>{@link ServletContext#getRealPath(String)}</li>
+     * <li>{@link ServletContext#getPath(String)}</li>
+     * <li>{@link ServletContext#getResource(String)}</li>
+     * <li>{@link ServletContext#getResourceAsStream(String)}</li>
+     * <li>{@link ServletContext#getResourceAsPath(String)}</li>
+     * <li>{@link ServletContext#getResourcePaths(String)}</li>
+     * <li>{@link AsyncContext#dispatch(String)}</li>
+     * <li>{@link AsyncContext#dispatch(ServletContext, String)}</li>
+     * </ul>
+     * for the purposes of these modes:
+     * <ul>
+     * <li>a 'reserved character' is a character defined as <code>reserved</code> by
+     * <a href="https://datatracker.ietf.org/doc/html/rfc3986#section-2.2">RFC3986 Section 2.2</a> or the ';'
+     * character.</li>
+     * <li>a 'path parameter' is the part of a URI segment as defined by the obsoleted
+     * <a href="https://datatracker.ietf.org/doc/html/rfc2396#section-3.3">RFC2396 Section 3.3</a>.</li>
+     * </ul>
+     * 
+     * @since Servlet 6.0
+     */
+    public enum PathCodingMode {
+        /**
+         * Effected methods will return/accept paths as was done prior to the introduction of the PathCodingMode in Servlet 6.0.
+         * The path returned by {@link HttpServletRequest#getContextPath()} is not decoded. Paths returned by the methods
+         * {@link HttpServletRequest#getServletPath()} and {@link HttpServletRequest#getPathInfo()} are fully decoded and have
+         * all path parameters removed. Decoding may include ambiguous characters such as a decoded <code>%2F</code>, which is
+         * indistinguishable from a segment separator '/' in the request URI. Methods accepting paths may accept encoded
+         * characters. A '?' character in a path will be treated as indicating a query string and will not be encoded. Some
+         * implementation specific behavioural differences may be apparent in this mode.
+         */
+        LEGACY,
+
+        /**
+         * Effected methods will only return/accept paths that do not contain any encoded reserved characters and with all path
+         * parameters removed. If the requested URI contains encoded reserved characters, then a call to a method that returns
+         * such a path will throw {@link IllegalStateException}.
+         *
+         * <p>
+         * TODO pick between this: If a path with encoded reserve characters is passed to a method taking a path as a parameter,
+         * then it will throw {@link MalformedURLException} if declared. otherwise a {@link IllegalArgumentException}.
+         *
+         * <p>
+         * TODO or this: If a path with encoded reserve characters is received, then the request will be rejected by the servlet
+         * container with a 400 BAD REQUEST response.
+         */
+        DECODED,
```

Comment:
I think this can be simplified. Once path parameters have been removed, it is safe to decode any %nn reserved characters apart from `%2F`. Therefore, I think we can remove the requirement to throw an error in this mode entirely. This mode then becomes:
- decode unreserved %nn
- remove all path parameters
- normalize
- decode reserved %nn except `%2F`

This achieves the same result without causing problems for apps (like Jira) that use encoded reserved characters in some requests and removes the need to triggering exceptions that will break some apps that would otherwise work correctly.

Note: There is an interesting edge case here. `/aaa/..;a=b/`. The proposal above would result in `/`. However it can be argued that this is inconsistent with RFC 3986. There are other approaches. Given there is no legitimate reason I can think of for sequences like this I quite like rejecting such requests with a 400 response.

Reply:
The beauty of what you propose here is that it get's us really close to not needing to be modal, as we would only ever return sane paths here and insane ones would be rejected with 400.  

But I'm not so sure that %2F is really the only problem character?  

Potentially there are also issues with URIs like `/aaa/..%5c../foo` which according to RFC3986 should decode to `/aaa/..\../foo`, which can be missinterpreted on some file systems.    

We've also seen issues with URIs like `/foo%00`.  Ah but this indicates a problem with specifing in terms of [reserved](https://datatracker.ietf.org/doc/html/rfc3986#section-2.3) from the RFC.  I think we probably should specify in terms of [unreserved](https://datatracker.ietf.org/doc/html/rfc3986#section-2.3), as there is a gap between those definitions that includes control characters.

Also if we allow encoded '%' (%25), then the resulting paths can contain percent sequences that can be dangerous to pass onto methods that a little flexible about receiving encoded characters or not.  So for example `/foo/%252E%252E/bar` could result in a `/foo/%2e%2e/bar` path that may then get later converted to `/foo/../bar` and normalized to just `/bar`

So I'm not totally convinced that %2F is the only danger encoding.

Regarding the edge case of `/aaa/..;`, jetty is now rejecting such requests with 400 unless told to do otherwise.
The question is, if we do let such requests through, should the result be `/` or `/aaa/..` ?  The later is compliant with RFC3986 but potentially dangerous when passed to file handling code.   I think your process about would result in `/`

So to be compliant with RFC3986 we'd could have a process like:
 + decode unreserved %nn
 + normalize
 + remove all path parameters
 + reject ambigous/scary/insane URIs. eg containing %2F, ".." etc.  See [Jetty's enumeration of scary URIs](https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java).
 + decode reserved %nn 

But despite these reservations, I think the general approach of path methodes never returning scary paths may be workable.   But my concern is that we then move the "modes" to the definition of what is scary.  Is that something we need to standardize? I think we need to at least enumerate the ones that should be rejected by default... we could then perhaps leave it up to containers to configure exceptions.    

However, if we want applications that handle the exceptional cases by calling getRequestURI to be portable, then maybe we need to only fail scary URIs if the path methods are called rather than reject up front?

I also think we need to think through the modes on how methods that accept paths act.  Currently I think there is a lot of flexibility, which is why so many scary URIs can be damaging if let flow through to these methods.    I'm not sure how we can make them stricter without introducing at least a LEGACY mode.

Geeze this is a horrid problem!